File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.cpp
// class Solution {
// public:
//     int maxIceCream(vector<int>& costs, int coins) {
//         sort(costs.begin(), costs.end());
//         int i=0;
//         while (i<costs.size() && coins>=costs[i])
//             coins-=costs[i++];
//         return i;
//     }
// };

class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        vector<int> map(100001, 0);
        int i, res=0;
        for(auto c: costs)
            map[c]++;
        for(i=1; i<=min(100000, coins); i++){
            int canBuy = min(map[i], coins/i);
            coins-=canBuy*i;
            res+=canBuy;
        }
        return res;
    }
};

File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.java
class Solution {
    public int maxIceCream(int[] costs, int coins) {
        int maxm = 0, ans = 0;
        for(int i: costs)
            maxm = Math.max(maxm, i);
        int[] counts = new int[maxm+1];
        for(int i: costs)
            counts[i]++;
        for(int i=0; i<counts.length; i++){
            if(coins > counts[i] * i){
                coins -= counts[i] * i;
                ans += counts[i];
            } else {
                ans += coins / i;
                break;
            }
        }
        return ans;
    }
}